File: 0200-number-of-islands/0200-number-of-islands.js
/**
 * @param {character[][]} grid
 * @return {number}
 */
var numIslands = function(grid) {
    let islands = 0;
    let mRows = grid.length;
    let nColumns = grid[0].length;
    let direction = [[0, -1], [1, 0], [0, 1], [-1, 0]];
    
    const dfs = function(i, j) { 
        grid[i][j] = 2;
        let r, c;
        for (let dir of direction) {
            r = i + dir[0];
            c = j + dir[1];
            if (r >= 0 && r < mRows && c >= 0 && c < nColumns && grid[r][c] == 1) {
                dfs(r, c);
            }
        }
        return;
    };
    
    for (let i = 0; i < mRows; i++) {
        for (let j = 0; j < nColumns; j++) {
            if (grid[i][j] == 1) {
                dfs(i, j);
                islands++;
            }            
        }
    }
    return islands;   
};

File: 0200-number-of-islands/0200-number-of-islands.py
class Solution(object):
    def numIslands(self, grid):
        islands = 0
        mRows = len(grid)
        nColumns = len(grid[0])
        direction = [[0, -1], [1, 0], [0, 1], [-1, 0]]
        
        def dfs(grid, i, j):
            grid[i][j] = '0'
            for dir in range(len(direction)):
                r = i + direction[dir][0]
                c = j + direction[dir][1]
                if r < 0 or c < 0 or r >= mRows or c >= nColumns or grid[r][c] == '0':
                    continue
                dfs(grid, r, c)
            return
        
        for i in range(mRows):
            for j in range(nColumns):
                if grid[i][j] == '1':
                    dfs(grid, i, j)
                    islands += 1
                    
        return islands

        